File: shared/protocol/csharp/Protocol.cs
// SecondScreen Local — Wire Protocol v1 (C# mirror)
// Canonical spec: /shared/protocol/PROTOCOL.md. Keep the three language mirrors in sync.
// NOTE: namespace is SecondScreen.Core (not .Shared) so the C# Core project — which has
// ImplicitUsings and file-scoped `namespace SecondScreen.Core` everywhere — resolves these
// symbols without an extra using. The Desktop project's `using SecondScreen.Core;` also covers
// QualityMode/SessionState.
namespace SecondScreen.Core
{
    public static class Protocol
    {
        public const int Version = 1;

        // Ports (see PROTOCOL.md §1)
        public const int DiscoveryUdpPort = 47800;
        public const int ControlTcpPort   = 47801;
        public const int VideoUdpPort      = 47802;

        // Discovery payload types
        public const string DiscoverType = "SSL_DISCOVER";
        public const string AnnounceType = "SSL_ANNOUNCE";

        // Crypto parameters (PROTOCOL.md §3.2)
        public const string EcdhCurve   = "nistP256"; // P-256 / secp256r1
        public const string HkdfInfo    = "SecondScreenLocal/v1/session";
        public const int    SessionKeyBytes = 32;      // AES-256
        public const int    GcmNonceBytes   = 12;
        public const int    GcmTagBytes     = 16;
        public const int    PinDigits       = 6;
        public const int    MaxPairFailures = 5;

        // Timeouts / heartbeat
        public const int HeartbeatIntervalMs = 1000;
        public const int MissedPongsToDrop   = 3;
        public const int ReconnectMaxAttempts = 5;
        public const int ReconnectGraceMs     = 15000;

        // Video packet header (PROTOCOL.md §5)
        public static readonly byte[] VideoMagic = { 0x53, 0x56 }; // 'S','V'
        public const int VideoHeaderBytes = 22;
        public const int VideoMaxPayload  = 1200;
        // flags
        public const byte FlagKeyframe   = 0x01;
        public const byte FlagLastPacket = 0x02;
        public const byte FlagEncrypted  = 0x04;
    }

    // Control channel JSON "type" values (PROTOCOL.md §3)
    public static class MessageType
    {
        public const string Hello           = "HELLO";
        public const string HelloAck         = "HELLO_ACK";
        public const string PairConfirm      = "PAIR_CONFIRM";
        public const string PairOk           = "PAIR_OK";
        public const string PairFail         = "PAIR_FAIL";
        public const string SessionConfig    = "SESSION_CONFIG";
        public const string SessionConfigAck = "SESSION_CONFIG_ACK";
        public const string Ping             = "PING";
        public const string Pong             = "PONG";
        public const string Touch            = "TOUCH";
        public const string RequestKeyframe  = "REQUEST_KEYFRAME";
        public const string Stats            = "STATS";
        public const string DeviceUpdate     = "DEVICE_UPDATE";
        public const string SetQuality       = "SET_QUALITY";
        public const string Disconnect       = "DISCONNECT";
    }

    // Touch event kinds (PROTOCOL.md §4)
    public static class TouchEvent
    {
        public const string Down      = "DOWN";
        public const string Move      = "MOVE";
        public const string Up        = "UP";
        public const string Scroll    = "SCROLL";
        public const string LongPress = "LONG_PRESS";
    }

    public enum SessionState
    {
        Idle, Discovering, Connecting, Pairing, Configuring, Streaming, Reconnecting, Disconnected
    }

    public enum QualityMode { Performance, Balanced, HighQuality }
}

File: shared/protocol/kotlin/Protocol.kt
// SecondScreen Local — Wire Protocol v1 (Kotlin mirror)
// Canonical spec: /shared/protocol/PROTOCOL.md. Keep the three language mirrors in sync.
package com.secondscreen.local.shared

object Protocol {
    const val VERSION = 1

    // Ports (PROTOCOL.md §1)
    const val DISCOVERY_UDP_PORT = 47800
    const val CONTROL_TCP_PORT = 47801
    const val VIDEO_UDP_PORT = 47802

    // Discovery payload types
    const val DISCOVER_TYPE = "SSL_DISCOVER"
    const val ANNOUNCE_TYPE = "SSL_ANNOUNCE"

    // Crypto (PROTOCOL.md §3.2)
    const val EC_CURVE = "secp256r1"          // P-256
    const val EC_ALGORITHM = "EC"
    const val KEY_AGREEMENT = "ECDH"
    const val HKDF_INFO = "SecondScreenLocal/v1/session"
    const val SESSION_KEY_BYTES = 32          // AES-256
    const val GCM_NONCE_BYTES = 12
    const val GCM_TAG_BITS = 128
    const val PIN_DIGITS = 6

    // Heartbeat / reconnect
    const val HEARTBEAT_INTERVAL_MS = 1000L
    const val MISSED_PONGS_TO_DROP = 3
    const val RECONNECT_MAX_ATTEMPTS = 5
    const val RECONNECT_GRACE_MS = 15000L

    // Video packet header (PROTOCOL.md §5)
    val VIDEO_MAGIC = byteArrayOf(0x53, 0x56) // 'S','V'
    const val VIDEO_HEADER_BYTES = 22
    const val VIDEO_MAX_PAYLOAD = 1200

    // flags
    const val FLAG_KEYFRAME: Int = 0x01
    const val FLAG_LAST_PACKET: Int = 0x02
    const val FLAG_ENCRYPTED: Int = 0x04
}

object MessageType {
    const val HELLO = "HELLO"
    const val HELLO_ACK = "HELLO_ACK"
    const val PAIR_CONFIRM = "PAIR_CONFIRM"
    const val PAIR_OK = "PAIR_OK"
    const val PAIR_FAIL = "PAIR_FAIL"
    const val SESSION_CONFIG = "SESSION_CONFIG"
    const val SESSION_CONFIG_ACK = "SESSION_CONFIG_ACK"
    const val PING = "PING"
    const val PONG = "PONG"
    const val TOUCH = "TOUCH"
    const val REQUEST_KEYFRAME = "REQUEST_KEYFRAME"
    const val STATS = "STATS"
    const val DEVICE_UPDATE = "DEVICE_UPDATE"
    const val SET_QUALITY = "SET_QUALITY"
    const val DISCONNECT = "DISCONNECT"
}

object TouchEvent {
    const val DOWN = "DOWN"
    const val MOVE = "MOVE"
    const val UP = "UP"
    const val SCROLL = "SCROLL"
    const val LONG_PRESS = "LONG_PRESS"
}
